zkevm-circuits/src/evm_circuit/execution/create.rs

```diff
@@ -0,0 +1,924 @@
+use crate::{
+    evm_circuit::{
+        execution::ExecutionGadget,
+        param::{
+            N_BYTES_ACCOUNT_ADDRESS, N_BYTES_GAS, N_BYTES_MEMORY_ADDRESS, N_BYTES_MEMORY_WORD_SIZE,
+            N_BYTES_U64, N_BYTES_WORD,
+        },
+        step::ExecutionState,
+        util::{
+            common_gadget::TransferGadget,
+            constraint_builder::{
+                ConstrainBuilderCommon, EVMConstraintBuilder, ReversionInfo, StepStateTransition,
+                Transition::{Delta, To},
+            },
+            math_gadget::{
+                ConstantDivisionGadget, ContractCreateGadget, IsZeroGadget, LtGadget, LtWordGadget,
+            },
+            memory_gadget::{MemoryAddressGadget, MemoryExpansionGadget},
+            not, CachedRegion, Cell, Word,
+        },
+        witness::{Block, Call, ExecStep, Transaction},
+    },
+    table::{AccountFieldTag, CallContextFieldTag},
+    util::Expr,
+};
+use bus_mapping::{circuit_input_builder::CopyDataType, evm::OpcodeId, state_db::CodeDB};
+use eth_types::{
+    evm_types::{GasCost, INIT_CODE_WORD_GAS},
+    Field, ToBigEndian, ToLittleEndian, ToScalar, U256,
+};
+use ethers_core::utils::keccak256;
+use gadgets::util::{and, expr_from_bytes, or, select};
+use halo2_proofs::{circuit::Value, plonk::Error};
+
+use std::iter::once;
+
+/// Gadget for CREATE and CREATE2 opcodes
+#[derive(Clone, Debug)]
+pub(crate) struct CreateGadget<F, const IS_CREATE2: bool, const S: ExecutionState> {
+    opcode: Cell<F>,
+    tx_id: Cell<F>,
+    reversion_info: ReversionInfo<F>,
+    depth: Cell<F>,
+
+    is_create2: IsZeroGadget<F>,
+    is_success: Cell<F>,
+    was_warm: Cell<F>,
+    value: Word<F>,
+
+    caller_balance: Word<F>,
+    callee_reversion_info: ReversionInfo<F>,
+    callee_nonce: Cell<F>,
+    prev_code_hash: Cell<F>,
+    transfer: TransferGadget<F>,
+    create: ContractCreateGadget<F, IS_CREATE2>,
+
+    init_code: MemoryAddressGadget<F>,
```

Comment:
We may add a `init_code_size_not_overflow: LtGadget<F, { N_BYTES_MEMORY_ADDRESS }>` to make `init_code.length() < 49153`.

It is part of shanghai EIP-3860, but it could not break CI tests, I think we may add it in a separate PR.

Could reference [gasCreateEip3860](https://github.com/ethereum/go-ethereum/blob/master/core/vm/gas_table.go#LL311C7-L311C7).

Reply:
issue created #1458